app/src/main/java/com/github/se/polyfit/model/ingredient/Ingredient.kt

```diff
@@ -0,0 +1,74 @@
+package com.github.se.polyfit.model.ingredient
+
+import com.github.se.polyfit.model.meal.Meal
+import kotlin.properties.Delegates
+
+/**
+ * Represents an ingredient in a meal. Automatically updates the parent meal when any property
+ * changes.
+ *
+ * @property name is unique for each meal
+ */
+class Ingredient(
+    val name: String, // name is now a val
+    servingSize: Double,
+    calories: Double,
+    protein: Double,
+    carbohydrates: Double,
+    fat: Double
```

Comment:
Ingredients will likely have a lot more nutritional information (see this [design](https://www.figma.com/file/JAs7WfDzS8lFij1xYd2J1I/User-Stories?type=design&node-id=60-515&mode=design&t=2aFeg5xv7Ml8Dtfl-4) for example. I'm wondering thoughts on creating a `NutritionalInformation` class, or something similar? Also, we might have different size formats (e.g. ml, g, 1 unit (i.e. for an oreo for example), so it may be worth creating an enum of sizeTypes, and maybe also a map for converting between similar ones (e.g. servingSize = 0.2 and sizeType = g, so we know this is 200mg). Happy to discuss more just initial thoughts

Reply:
Exactly what I was thinking, I tried an implementation along those lines. But in order to implement something like that, the auto update feature of the meal when an new ingredient is added (at least the current implementation).

Did you already have an interface/abstract class design in mind ?

Reply:
Additionally, the issue with the API I have in mind (LogMeal) is that the documentation lacks detail. It does not describe which nutritional values are returned. This is why it is modeled in a simplistic way. I am waiting for @Sobell13 to provide more information about that.

Reply:
Here is a sample response from the most likely API we will use: 
{
    "foodName": [
        "meat lasagna"
    ],
    "hasNutritionalInfo": true,
    "ids": [
        1034
    ],
    "imageId": 1368311,
    "nutritional_info": {
        "calories": 602.5,
        "totalNutrients": {
            "CA": {
                "label": "Calcium",
                "quantity": 527.01,
                "unit": "mg"
            },
            "CHOCDF": {
                "label": "Carbs",
                "quantity": 34.72,
                "unit": "g"
            },
            "CHOLE": {
                "label": "Cholesterol",
                "quantity": 165.85,
                "unit": "mg"
            },
            "ENERC_KCAL": {
                "label": "Energy",
                "quantity": 602.5,
                "unit": "kcal"
            },
            "FAMS": {
                "label": "Monounsaturated fats",
                "quantity": 10.67,
                "unit": "g"
            },
            "FAPU": {
                "label": "Polyunsaturated",
                "quantity": 2.57,
                "unit": "g"
            },
            "FASAT": {
                "label": "Saturated",
                "quantity": 15.29,
                "unit": "g"
            },
            "FAT": {
                "label": "Fat",
                "quantity": 32.45,
                "unit": "g"
            },
            "FATRN": {
                "label": "Trans fat",
                "quantity": 0.31,
                "unit": "g"
            },
            "FE": {
                "label": "Iron",
                "quantity": 4.6,
                "unit": "mg"
            },
            "FIBTG": {
                "label": "Fiber",
                "quantity": 3.92,
                "unit": "g"
            },
            "FOLAC": {
                "label": "Folic acid",
                "quantity": 34.24,
                "unit": "µg"
            },
            "FOLDFE": {
                "label": "Folate equivalent (total)",
                "quantity": 100.81,
                "unit": "µg"
            },
            "FOLFD": {
                "label": "Folate (food)",
                "quantity": 42.7,
                "unit": "µg"
            },
            "K": {
                "label": "Potassium",
                "quantity": 1041.44,
                "unit": "mg"
            },
            "MG": {
                "label": "Magnesium",
                "quantity": 75.08,
                "unit": "mg"
            },
            "NA": {
                "label": "Sodium",
                "quantity": 1575.92,
                "unit": "mg"
            },
            "NIA": {
                "label": "Niacin (B3)",
                "quantity": 6.26,
                "unit": "mg"
            },
            "P": {
                "label": "Phosphorus",
                "quantity": 609.72,
                "unit": "mg"
            },
            "PROCNT": {
                "label": "Protein",
                "quantity": 43.91,
                "unit": "g"
            },
            "RIBF": {
                "label": "Riboflavin (B2)",
                "quantity": 0.69,
                "unit": "mg"
            },
            "SUGAR": {
                "label": "Sugars",
                "quantity": 11.27,
                "unit": "g"
            },
            "SUGAR.added": {
                "label": "Sugars, added",
                "quantity": 0.0,
                "unit": "g"
            },
            "THIA": {
                "label": "Thiamin (B1)",
                "quantity": 0.3,
                "unit": "mg"
            },
            "TOCPHA": {
                "label": "Vitamin E",
                "quantity": 3.48,
                "unit": "mg"
            },
            "VITA_RAE": {
                "label": "Vitamin A",
                "quantity": 246.13,
                "unit": "µg"
            },
            "VITB12": {
                "label": "Vitamin B12",
                "quantity": 2.68,
                "unit": "µg"
            },
            "VITB6A": {
                "label": "Vitamin B6",
                "quantity": 0.47,
                "unit": "mg"
            },
            "VITC": {
                "label": "Vitamin C",
                "quantity": 14.78,
                "unit": "mg"
            },
            "VITD": {
                "label": "Vitamin D",
                "quantity": 40.8,
                "unit": "µg"
            },
            "VITK1": {
                "label": "Vitamin K",
                "quantity": 16.54,
                "unit": "µg"
            },
            "ZN": {
                "label": "Zinc",
                "quantity": 5.82,
                "unit": "mg"
            }
        }
    },
    "serving_size": 385.2
}

Reply:
Here's an idea in pseudocode but not 100% sure:

```kt
enum class MeasurementUnit {
    Kilogram, // could also do gram and milliliter, but i think best to keep this same for math simplicity later
    Litre,
    Unit // For standard sized items, i.e. 1 oreo, 1 lollipop
}

class Ingredient (
  val name: String,
  val unitType: MeasurementUnit,
  // All of these values should be normalized to the unit type. So if we know something is 10 calories for 100ml, then this should be 100 calories since it is liters.
  val calories: Double,
  val carbohydrates: Double,
  // Include all nutritional info we want to store here
) { ... }

class Serving (
  val ingredient: Ingredient,
  val servingSize: Double // This will be in terms of base units. So, if someone has 100ml then its 0.1 for litres
  val meal: Meal
) { ... }

class Meal (
  val occasion: MealOccasion,
  val name: String,
  private val servings: MutableList<Serving> = mutableListOf()
) { ... }
```

and then we can maybe have something calculate the totals for meal?

Reply:
I think we can set it up for now with ingredients just having basic macros and add more after we know more about the API from alex

Reply:
The nice thing about structuring the ingredients this way is we can always check our DB for the ingredient before we send the API request out, so that eventually we will build our own database of nutritional info

Reply:
In order to get the ingredients that is another API call , but the unit for the response is just weight (haha). So I am guessing grames ??
Here is the structured data for a dish, specifically "meat lasagna":

```json
{
    "dish_id": [1034],
    "foodName": ["meat lasagna"],
    "imageId": 1368311,
    "recipe": [
        {
            "id": 522,
            "name": "beef",
            "weight": 30.74
        },
        {
            "id": 807,
            "name": "pork sausage",
            "weight": 22.94
        },
        {
            "id": 750,
            "name": "tomato sauce",
            "weight": 74.34
        },
        {
            "id": 112,
            "name": "tomato",
            "weight": 39.75
        },
        {
            "id": 381,
            "name": "garlic",
            "weight": 0.61
        },
        {
            "id": 47,
            "name": "sugar",
            "weight": 0.84
        },
        {
            "id": 925,
            "name": "italian seasoning",
            "weight": 0.18
        },
        {
            "id": 45,
            "name": "salt",
            "weight": 0.61
        },
        {
            "id": 175,
            "name": "pepper",
            "weight": 0.12
        },
        {
            "id": 569,
            "name": "egg",
            "weight": 15.17
        },
        {
            "id": 256,
            "name": "parsley",
            "weight": 0.49
        },
        {
            "id": 862,
            "name": "cottage cheese",
            "weight": 65.99
        },
        {
            "id": 837,
            "name": "ricotta cheese",
            "weight": 24.88
        },
        {
            "id": 18,
            "name": "parmesan cheese",
            "weight": 5.06
        },
        {
            "id": 926,
            "name": "lasagna noodles",
            "weight": 51.88
        },
        {
            "id": 927,
            "name": "provolone cheese",
            "weight": 17.2
        },
        {
            "id": 97,
            "name": "mozzarella",
            "weight": 34.41
        }
    ],
    "serving_size": 385.2
}

Reply:
can you link me the docs? i agree this is a weird response body lol

Reply:
So to resume this huge spam: For meals, we get loads of nutrition facts on nutrients and macros, but with all kinds of different measurements. As for ingredients, the API just gives us their weight and that's about it—no clue on what's actually inside them.

But hey, there's another cool API over at https://spoonacular.com/food-api/docs#Get-Recipe-Information that can break down the ingredients if you throw a recipe at it.

Reply:
Yeah, this API is much better. It gives you a breakdown of the recipe's nutritional info and other details, including the nutritional information about the ingredients (https://spoonacular.com/food-api/docs#Nutrition-by-ID).

It has an API for recipes and more. The only downside is that its image recognition is limited. But I think that's okay for the scope of this project. What do you think, @medwardson?